firefox-ios/Storage/RustKeychain.swift

```diff
@@ -0,0 +1,68 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Common
+
+open class RustKeychain {
+    public private(set) var serviceName: String
+    public private(set) var accessGroup: String?
+
+    static var sharedClientAppContainerKeychain: RustKeychain {
+        let baseBundleIdentifier = AppInfo.baseBundleIdentifier
+
+        guard let accessGroupPrefix = Bundle.main.object(forInfoDictionaryKey: "MozDevelopmentTeam") as? String else {
+            return RustKeychain(serviceName: baseBundleIdentifier)
+        }
+        return RustKeychain(serviceName: baseBundleIdentifier,
+                            accessGroup: AppInfo.keychainAccessGroupWithPrefix(accessGroupPrefix))
+    }
+
+    public init(serviceName: String, accessGroup: String? = nil) {
+        self.serviceName = serviceName
+        self.accessGroup = accessGroup
+    }
+
+    struct KeychainError: Error {
+        let errorMessage: String
+    }
+
+    func getBaseKeychainQuery(key: String) -> [String: Any?] {
+        let encodedIdentifier: Data? = key.data(using: String.Encoding.utf8)
+        let keychainQueryDictionary: [String: Any?] = [kSecClass as String: kSecClassGenericPassword,
+                                                       kSecAttrAccessGroup as String: self.accessGroup,
+                                                       kSecAttrService as String: self.serviceName,
+                                                       kSecAttrGeneric as String: encodedIdentifier,
+                                                       kSecAttrAccount as String: encodedIdentifier,
+                                                       kSecAttrSynchronizable as String: false]
+        return keychainQueryDictionary
+    }
+
+    func queryKeychainForKey(key: String) -> Result<Data?, Error> {
+        var keychainQueryDictionary = getBaseKeychainQuery(key: key)
+        keychainQueryDictionary[kSecMatchLimit as String] = kSecMatchLimitOne
+        keychainQueryDictionary[kSecReturnData as String] = kCFBooleanTrue
+
+        var queryResult: AnyObject?
+        let status = SecItemCopyMatching(keychainQueryDictionary as CFDictionary, &queryResult)
+
+        guard status == noErr else {
+            let errMsg = SecCopyErrorMessageString(status, nil)
+            return .failure(KeychainError(errorMessage: errMsg as? String ?? ""))
+        }
+        return .success(queryResult as? Data)
+    }
+
+    func updateKeychainKey(_ data: Data, key: String) -> OSStatus {
+        return SecItemUpdate(getBaseKeychainQuery(key: key) as CFDictionary,
```

Comment:
I know we're doing as 1:1 thing here but does it make sense to combine `update` and `set` that way we can prevent certain issues (like adding a new key that already exists, etc)?

Reply:
Might've highlighted the wrong line here, was mentioning more of `updateKeychainKey` and `setKeychainKey` combined into like a `addOrUpdateKeychainKey` that way the callers don't need to worry about if a key already exists.

Reply:
Nope, I completely misread your original comment--my bad! I was too far down a rabbit hole. Anyway that's a fair point because `MZKeychainWrapper`'s set function is an upsert and we don't ever call update. It'll also make my calling code cleaner. Let me see what I can do.

Reply:
And here's [a commit](https://github.com/mozilla-mobile/firefox-ios/pull/24844/commits/b818dcee3d0753a1b7f5e31e6c49e6a771511461) for you